pkg/server/commands/reverseexpand/reverse_expand.go

```diff
@@ -639,3 +657,125 @@ func (c *ReverseExpandQuery) throttle(ctx context.Context, currentNumDispatch ui
+//
+// For BUT NOT relations, getEdgesFromWeightedGraph first checks if the BUT NOT applies to the source type, and if it
+// does it will mark this result as "requires check".
+// E.g. If we have `rel1: a OR b BUT NOT c` and we are searching for a "user", if 'c' does not lead to type user,
+// we do not mark as "requires check".
+// After determining whether this result will require check, getEdgesFromWeightedGraph will prune off the last edge of the
+// Exclusion, as the right-most edge is always the BUT NOT portion, and that edge has already been accounted for.
+//
+// getEdgesFromWeightedGraph returns a list of edges, boolean indicating whether Check is needed, and an error.
+func (c *ReverseExpandQuery) getEdgesFromWeightedGraph(
+	wg *weightedGraph.WeightedAuthorizationModelGraph,
+	targetTypeRelation string,
+	sourceType string,
+	needsCheck bool,
+) ([]*weightedGraph.WeightedAuthorizationModelEdge, bool, error) {
+	if wg == nil {
+		// this should never happen
+		return nil, false, errors.New("weighted graph is nil")
+	}
+
+	currentNode, ok := wg.GetNodeByID(targetTypeRelation)
+	if !ok {
+		// This should never happen
+		return nil, false, errors.New("currentNode is nil")
+	}
+
+	// This means we cannot reach the source type requested.
+	// e.g. there is no path from 'document' to 'user'
+	if _, ok = currentNode.GetWeight(sourceType); !ok {
+		return nil, false, nil
+	}
+
+	edges, _ := wg.GetEdgesFromNode(currentNode)
+
+	// TODO: this _shouldn't_ be reachable but will be dealt with in a follow up PR
+	// This would mean that we dispatched from a direct edge, which doesn't make sense
+	if len(edges) == 0 {
+		return nil, false, errors.New("no outgoing edges")
+	}
+
+	if currentNode.GetNodeType() == weightedGraph.OperatorNode {
+		switch currentNode.GetLabel() {
+		case weightedGraph.ExclusionOperator: // e.g. rel1: [user, other] BUT NOT b
+			butNotEdge := edges[len(edges)-1] // this is the edge to 'b'
+			_, canReachSource := butNotEdge.GetWeight(sourceType)
+
+			// if the 'b' in BUT NOT b has a weight for the terminal type we're seeking
+			// we need to run check at the end
+			if canReachSource {
+				needsCheck = true
+			}
+
+			// prune off the "BUT NOT b" portion of these edges and keep going
+			// the right-most edge is ALWAYS the "BUT NOT", so trim the last element
+			edges = edges[:len(edges)-1]
+		case weightedGraph.IntersectionOperator:
+			// For AND relations, mark as "needs check" and just pick the lowest weight edge
+			needsCheck = true
+
+			lowestWeightEdge := reduce(edges, nil, cheapestEdgeTo(sourceType))
+
+			// return only the lowest weight edge
+			edges = []*weightedGraph.WeightedAuthorizationModelEdge{lowestWeightEdge}
+		}
+	}
+
+	// Filter to only return edges which have a path to the sourceType
+	relevantEdges := filter(edges, hasPathTo(sourceType))
+
+	return relevantEdges, needsCheck, nil
+}
+
+func hasPathTo(dest string) func(*weightedGraph.WeightedAuthorizationModelEdge) bool {
```

Comment:
Thus def should be in the typesystem (there is a path exists which this should replace it altogether).